File: uninformed search/bidirectional_search.cpp
#include<bits/stdc++.h>
using namespace std;

vector <int> adj[1000];
int n,e;

bool visited_forward[1000];
bool visited_backward[1000];

int pf[1000],pb[1000];

void path(int u,int s)
{
    if(u==s)
    {
        cout<<u<<" ";
        return;
    }

    path(pf[u],s);
    cout<<u<<" ";
}

void path2(int s,int d)
{
    while(s!=d)
    {
        s=pb[s];
        cout<<s<<" ";
    }
}




void bidirection(int src,int dest)
{

    queue <int> qf,qb;

    visited_forward[src]=true;
    visited_backward[dest]=true;

    //level[src]=0;
    //p[src]=-1;

    qf.push(src);
    qb.push(dest);

    while(!qf.empty() && !qb.empty())
    {
        ///forward bfs

        int x=qf.front();
        qf.pop();
        cout<<x<<" for"<<endl;

        for(int i=0; i<adj[x].size(); i++)
        {
            int y=adj[x][i];
            if(visited_forward[y]==false)
            {
                visited_forward[y]=true;
                pf[y]=x;
                qf.push(y);

                if(visited_backward[y]==true)
                {
                    cout<<"Intersected at "<<y<<endl;
                    cout<<"Path : ";
                    path(y,src);
                    path2(y,dest);
                    return;
                }
            }


        }


        ///backward bfs
        x=qb.front();
        qb.pop();
        cout<<x<<" back"<<endl;

        for(int i=0; i<adj[x].size(); i++)
        {
            int y=adj[x][i];
            if(visited_backward[y]==false)
            {
                visited_backward[y]=true;
                pb[y]=x;
                qb.push(y);

                if(visited_forward[y]==true)
                {
                    cout<<"Intersected at -"<<y<<endl;
                    cout<<"Path : ";
                    path(y,src);
                    path2(y,dest);
                    return;
                }
            }


        }
    }



}




int main()
{

    //cout<<"Enter the number of nodes : ";
    cin>>n;
    //cout<<"Enter the number of edges : ";
    cin>>e;

    //cout<<"Enter "<<e<<" edges : ";

    int u,v;
    for(int i=0; i<e; i++)
    {
        cin>>u>>v;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    int src,dest;
    cin>>src>>dest;

    bidirection(src,dest);

    return 0;
}


/*
15 14
0 4
1 4
2 5
3 5
4 6
5 6
6 7
7 8
8 9
8 10
9 11
9 12
10 13
10 14
*/

File: Bidirectional_Search.py
from collections import defaultdict, deque


def printpath(Fpath,Bpath):
    
    Fpath.pop()
    Bpath.reverse()
    path=Fpath+Bpath

    print(path)


def bidirectional(adj,src,goal):
    
    visited_forward=[]
    visited_backward=[]
    visited_forward.append(src)
    visited_backward.append(goal)
    
    forward_queue=deque()
    backward_queue=deque()
    forward_queue.append(src)
    backward_queue.append(goal)
    
    forward_path=defaultdict(list)
    backward_path=defaultdict(list)
    
    forward_path[src].append(src)
    backward_path[goal].append(goal)
    
    intersection=-1
    
    while len(forward_queue) and len(backward_queue):
    
        #forward BFS
        x=forward_queue.popleft()
        print("forward :",x)
        
    
        for node in adj[x]:
            if node not in visited_forward:
                visited_forward.append(node)
                forward_path[node]=forward_path[x]+[node]
                
                if node in visited_backward:
                    intersection=node
                    print(intersection)
                    break
                    
                forward_queue.append(node)
                
        
        if intersection!=-1:
            printpath(forward_path[intersection],backward_path[intersection])
            return
               
                
        #backward BFS
        x=backward_queue.popleft()
        print("backward :",x)
        
    
        for node in adj[x]:
            if node not in visited_backward:
                visited_backward.append(node)
                backward_path[node]=backward_path[x]+[node]
                
                if node in visited_forward:
                    intersection=node
                    print(intersection)
                    break
                    
                backward_queue.append(node)
                
                
        if intersection!=-1:
            printpath(forward_path[intersection],backward_path[intersection])
            return
            
    print("Path does not exist")
    
    



n=int(input("Enter the number of nodes : "))

nodes=[]

print("Enter",n,"nodes")
for i in range(n):
    x=input("node"+str(i+1)+" : ")
    nodes.append(x)
    
e=int(input("Enter the number of edges : "))

adj=defaultdict(list)

for i in range(e):
    print("Edge",i+1," : ")
    v=input("Start Node : ");
    u=input("End Node : ");
    
    if u not in nodes or v not in nodes:
        print("Invalid Edge")
        quit()
    
    adj[v].append(u)
    adj[u].append(v)
    
#print(adj)
src=input("Enter the source node : ")
goal=input("Enter the goal node : ")

bidirectional(adj,src,goal)
    

 

'''
15
0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
14
0
4
1
4
2
5
3
5
4
6
5
6
6
7
7
8
8
9
8
10
9
11
9
12
10
13
10
14
'''


    
